text/3453-f16-and-f128.md

```diff
@@ -0,0 +1,134 @@
+- Feature Name: `f16_and_f128`
+- Start Date: 2023-07-02
+- RFC PR: [rust-lang/rfcs#3453](https://github.com/rust-lang/rfcs/pull/3453)
+- Rust Issue: [rust-lang/rfcs#2629](https://github.com/rust-lang/rfcs/issues/2629)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes adding new IEEE-compliant floating point types `f16` and `f128` into the core language and standard library. We will provide a soft float implementation for all targets, and use hardware support where possible.
+
+# Motivation
+[motivation]: #motivation
+
+The IEEE 754 standard defines many binary floating point formats. The most common of these types are the binary32 and binary64 formats, available in Rust as `f32` and `f64`. However, other formats are useful in various uncommon scenarios. The binary16 format is useful for situations where storage compactness is important and low precision is acceptable, such as HDR images, mesh quantization, and AI neural networks.[^1] The binary128 format is useful for situations where high precision is needed, such as scientific computing contexts.
+
+The proposal is to add `f16` and `f128` types in Rust to represent IEEE 754 binary16 and binary128 respectively. Having `f16` and `f128` types in the Rust language would make Rust an optimal environment for more advanced use cases. Unlike third-party crates, this enables the compiler to perform optimizations for hardware with native support for these types, allows defining literals for these types, and would provide one single canonical data type for these floats, making it easier to exchange data between libraries.
+
+This RFC does not have the goal of covering the entire IEEE 754 standard, since it does not include `f256` and the decimal-float types. This RFC also does not have the goal of adding existing platform-specific float types such as x86's 80-bit double-extended-precision. This RFC does not make a judgement of whether those types should be added in the future, such discussion can be left to a future RFC, but it is not the goal of this RFC.
```

Comment:
One of the biggest things called out in the meeting was that motivation could use some enhancement. Some assorted things that you may want to add:

- Somehow say that the downsides to adding these are small (since we are letting LLVM do the work), but that this can be extremely important for some niches
- Point out what exactly can and can't be done via external crates. E.g. can provide an implementation, difficult to manage because it needs to rely on handwritten assembly and linking external libraries
- Maybe enhance the "scientific computing contexts" statement with more specific examples. You could probably mention how Julia and NumPy support f128 operations
- That there are numerous use cases for something more precise than `f64` - in C this is typically `long double`, Rust doesn't have anything. So we both don't have a way to do this more precise math, nor do we have a way to interop with C
- `long double` in C is already `_Float128` on many targets, so a lot of C libraries are using f128. Not having something is blocking us from some use cases, e.g. https://github.com/rust-lang/rust/issues/74393
- @pvdrz has mentioned making that easier for bindgen a few times, bindgen's job getting easier is probably worth mentioning specifically
- Something from the  [C++ motivation](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1467r9.html#motivation) might work for f16

Some of this is or could be in rationale and alternatives too, that section usually expands upon motivation